packages/bsky/src/ingester/subscription.ts

```diff
@@ -0,0 +1,248 @@
+        )
+      }
+      if (!this.destroyed) {
+        await wait(1000 + jitter(500)) // wait then try to become leader
+      }
+    }
+  }
+
+  async destroy() {
+    this.destroyed = true
+    await this.cursorQueue.destroy()
+    this.leader.destroy(new DisconnectError())
+  }
+
+  async resume() {
+    this.destroyed = false
+    this.cursorQueue = new LatestQueue()
+    await this.run()
+  }
+
+  async getCursor(): Promise<number> {
+    const val = await this.ctx.redis.get(this.ns(CURSOR_KEY))
+    const state = val !== null ? strToInt(val) : 0
+    return state
+  }
+
+  async resetCursor(): Promise<void> {
+    await this.ctx.redis.del(this.ns(CURSOR_KEY))
+  }
+
+  private async setCursor(seq: number): Promise<void> {
+    await this.ctx.redis.set(this.ns(CURSOR_KEY), seq)
+  }
+
+  private getSubscription(opts: { signal: AbortSignal }) {
+    return new Subscription({
+      service: this.opts.service,
+      method: METHOD,
+      signal: opts.signal,
+      getParams: async () => {
+        const cursor = await this.getCursor()
+        return { cursor }
+      },
+      onReconnectError: (err, reconnects, initial) => {
+        subLogger.warn(
+          { err, reconnects, initial },
+          'ingester subscription reconnect',
+        )
+      },
+      validate: (value) => {
+        try {
+          return lexicons.assertValidXrpcMessage<Message>(METHOD, value)
+        } catch (err) {
+          subLogger.warn(
+            {
+              err,
+              seq: ifNumber(value?.['seq']),
+              repo: ifString(value?.['repo']),
+              commit: ifString(value?.['commit']?.toString()),
+              time: ifString(value?.['time']),
+              provider: this.opts.service,
+            },
+            'ingester subscription skipped invalid message',
+          )
+        }
+      },
+    })
+  }
+
+  // namespace redis keys
+  ns(key: string) {
+    return this.opts.namespace ? `${this.opts.namespace}:${key}` : key
+  }
+}
+
+function ifString(val: unknown): string | undefined {
+  return typeof val === 'string' ? val : undefined
+}
+
+function ifNumber(val: unknown): number | undefined {
+  return typeof val === 'number' ? val : undefined
+}
+
+function getMessageDetails(msg: Message):
+  | { info: message.Info | null }
+  | {
+      seq: number
+      repo: string
+      message: ProcessableMessage
+    } {
+  if (message.isCommit(msg)) {
+    return { seq: msg.seq, repo: msg.repo, message: msg }
+  } else if (message.isHandle(msg)) {
+    return { seq: msg.seq, repo: msg.did, message: msg }
+  } else if (message.isMigrate(msg)) {
+    return { seq: msg.seq, repo: msg.did, message: msg }
+  } else if (message.isTombstone(msg)) {
+    return { seq: msg.seq, repo: msg.did, message: msg }
+  } else if (message.isInfo(msg)) {
+    return { info: msg }
+  }
+  return { info: null }
+}
+
+function ui8ToBuffer(bytes: Uint8Array) {
+  return Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength)
```

Comment:
ah nice! this could be a good function for `@atproto/common` - i think i've run into similar in the past